test-plan/eventing/channel/control-plane/channel-lifecycle-conformance.md

```diff
@@ -0,0 +1,228 @@
+
+
+
+# Testing Channel Lifecycle Conformance: 
+
+We are going to be testing the previous paragraphs coming from the Knative Eventing Spec. To do this we will be creating a Channel, checking its immutable properties, checking its Ready status and then creating a Subscription that links to it by making a reference. We will also checking the Subscription status, as it depends on the Channel to be ready to work correctly. We will be also checking that the Channel is addressable by looking at the status conditions fields. Because this is a Control Plane test, we are not going to be sending Events to the components. 
+
+You can find the resources for running these tests inside the [control-plane/channel-lifecycle/](test-plan/eventing/channel/control-plane/) directory. 
+- A [Channel resource](test-plan/eventing/channel/control-plane/channel.yaml)
+- A [Subscription resource that references the Channel](test-plan/eventing/channel/control-plane/subscription.yaml)
+- A [Service resource that serves as deadletter sink and subscriber for the subscritpion](test-plan/eventing/channel/control-plane/service.yaml)
+
+
+## [Pre] Creating a Channel 
+
+```
+kubectl apply -f channel/control-plane/channel.yaml
+```
+
+
+## [Test] Immutability
+
+Check for default annotations, this should return the name of the selected implementation: 
+
+```
+kubectl get channel.messaging.knative.dev conformance-channel -o jsonpath='{.spec.channelTemplate.kind}'
+```
+
+Try to patch the annotation: `messaging.knative.dev/channel.spec.channelTemplate` to see if the resource mutates: 
+
+```
+kubectl patch channel.messaging.knative.dev conformance-channel --type merge -p '{"spec":{"channelTemplate":{"kind":"mutable"}}}'
+```
+
+You should get the following error: 
+```
+Error from server (BadRequest): admission webhook "validation.webhook.eventing.knative.dev" denied the request: validation failed: Immutable fields changed (-old +new): annotations
+{string}:
+	-: "InMemoryChannel" // or your channel implementation
+	+: "mutable"
+```
+
+Tested in eventing:
+- Not tested
+
+### [Output]
+
+```
+{
+  "test": "channel/control-plane/immutability-1"
+  "output": {
+    	"channel Implementation": "<CHANNEL IMPLEMENTATION>",
+	"expectedError": "<EXPECTED ERROR>"
+  }
+}
+```
+
+## [Test] Channel Readiness 
+
+Check for condition type `Ready` with status `True`: 
+
+```
+ kubectl get channel.messaging.knative.dev conformance-channel -ojsonpath="{.status.conditions[?(@.type == \"Ready\")].status}"
+```
+
+Tested in eventing:
+- https://github.com/knative/eventing/blob/main/test/rekt/features/channel/control_plane.go#L102
+- https://github.com/knative/eventing/blob/main/test/conformance/channel_status_test.go#L29
```

Comment:
I'm not so sure about the value of adding those links. This code is about testing Knative eventing for conformance. What we really want is a conformance tool that can run against the Knative eventing reference implementation.

Reply:
Yep, but here we can se how to implement this tests so they are easier to automatize. Maybe @salaboy can help me to convince you with this :p or we en up deciding to remove it

Reply:
@lionelvillard the links added by @gabo1208 are just highlighting which resources and behaviours are already covered inside the `rekt` tests. We can use that to measure how much coverage we have today compared with the manual test plans. I am happy to move those links to a separate document but I think that having someone looking at the existing tests is a good thing to avoid wasting time or redoing tests that are already done. Does this make sense?

Reply:
I'm fine adding those links. Should they point to main or a release?

Reply:
I think to release. The only tedious thing is having to change all the links on this files from release to release

Reply:
@gabo1208 yeah.. but it makes sense to point them to a fixed release.

Reply:
Done with this change cc @salaboy @lionelvillard